src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs

```diff
@@ -321,23 +326,29 @@ private static unsafe long SendFile(SafeSocketHandle socket, SafeFileHandle file
 
         private static unsafe int SysReceive(SafeSocketHandle socket, SocketFlags flags, IList<ArraySegment<byte>> buffers, byte[]? socketAddress, ref int socketAddressLen, out SocketFlags receivedFlags, out Interop.Error errno)
         {
-            int available = 0;
-            errno = Interop.Sys.GetBytesAvailable(socket, &available);
-            if (errno != Interop.Error.SUCCESS)
-            {
-                receivedFlags = 0;
-                return -1;
-            }
-            if (available == 0)
+            int maxBuffers = buffers.Count;
+            bool allocOnStack = maxBuffers <= IovStackThreshold;
+
+            // When there are many buffers, reduce the number of pinned buffers based on available bytes.
+            int available = int.MaxValue;
+            if (!allocOnStack)
             {
-                // Don't truncate iovecs.
-                available = int.MaxValue;
+                errno = Interop.Sys.GetBytesAvailable(socket, &available);
+                if (errno != Interop.Error.SUCCESS)
+                {
+                    receivedFlags = 0;
+                    return -1;
+                }
+                if (available == 0)
+                {
+                    // Don't truncate iovecs.
+                    available = int.MaxValue;
+                }
             }
 
             // Pin buffers and set up iovecs.
-            int maxBuffers = buffers.Count;
-            var handles = new GCHandle[maxBuffers];
-            var iovecs = new Interop.Sys.IOVector[maxBuffers];
+            Span<GCHandle> handles = allocOnStack ? stackalloc GCHandle[IovStackThreshold] : new GCHandle[maxBuffers];
```

Comment:
Since IovStackThreshold is so small, I wonder if it would be worth it to have a dedicated path for a small number of buffers and rather than using pinned GCHandles using fixed.  The latter would necessitate writing out a fixed statement for each input, which is why it would need to be a small number, but it generally also has less overhead, in particular it's close to free when a GC doesn't occur, whereas there's always overhead to creating and free'ing GCHandles.

We can start with what you have and subsequently experiment with that if it's deemed a good idea, e.g. if we think two or three buffers for example is the most common case.  Then again, this is all the synchronous code path, so maybe we needn't invest a ton in optimizing it further.

Reply:
Sounds like an interesting idea to try. The relative gain will be small due to the high cost of the syscall. Are there some cases where this optimization is performed already in corefx?